lib/daru/index.rb

```diff
@@ -392,4 +474,109 @@ def sparse_tuples
       }
     end
   end
-end
+
+  class CategoricalIndex < Index
+    
+    include Enumerable
+    
+    attr_reader :cat_hash, :array, :map_cat_int, :map_int_cat
+    
+    def initialize indexes
+      # Create a hash to map each category to positional indexes
+      @cat_hash = Hash.new []
+      # To link every instance to its category,
+      # it stores integer for every instance representing its category
+      @array = []
+      # Map each category to a unique integer for effective storage in @array
+      map_cat_int = {}
+      # Inverse mapping of @map
+      @map_int_cat = {}
+      
+      cat_count = 0
+      indexes.each_with_index do |index, pos|
+        unless map_cat_int.include? index
+          map_cat_int[index] = cat_count
+          @map_int_cat[cat_count] = index
+          cat_count += 1
+        end
+        @cat_hash[index] += [pos]
+        @array << map_cat_int[index]
+      end
+    end
+    
+    def dup
+      Daru::CategoricalIndex.new to_a
+    end
+    
+    def include? index
+      @cat_hash.include? index
+    end
+    
+    def pos *args
+      positions = args.map do |index|
+        if include? index
+          @cat_hash[index]
+        elsif index.is_a?(Numeric) && index < @array.size
+          index
+        else
+          raise IndexError, "#{index.inspect} is neither a valid category"\
+            " nor a valid position"
+        end
```

Comment:
It complains about "Prefer single-quoted strings". So, you can just replace this with
```ruby
         raise IndexError, "#{index.inspect} is neither a valid category"\
            ' nor a valid position'
```
(though I can agree it can look a bit strange.)